server/src/main/java/org/elasticsearch/cluster/coordination/ValidateJoinRequest.java

```diff
@@ -7,32 +7,96 @@
  */
 package org.elasticsearch.cluster.coordination;
 
+import org.elasticsearch.Version;
 import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.common.CheckedSupplier;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.compress.CompressorFactory;
+import org.elasticsearch.common.io.stream.InputStreamStreamInput;
+import org.elasticsearch.common.io.stream.NamedWriteableAwareStreamInput;
+import org.elasticsearch.common.io.stream.NamedWriteableRegistry;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.core.RefCounted;
+import org.elasticsearch.core.internal.io.IOUtils;
 import org.elasticsearch.transport.TransportRequest;
 
 import java.io.IOException;
 
 public class ValidateJoinRequest extends TransportRequest {
-    private ClusterState state;
+    private final CheckedSupplier<ClusterState, IOException> stateSupplier;
+    private final RefCounted refCounted;
 
     public ValidateJoinRequest(StreamInput in) throws IOException {
         super(in);
-        this.state = ClusterState.readFrom(in, null);
+        if (in.getVersion().onOrAfter(Version.V_8_3_0)) {
+            // recent versions send a BytesTransportRequest containing the compressed state
+            final var bytes = in.readReleasableBytesReference();
+            final var version = in.getVersion();
+            final var namedWriteableRegistry = in.namedWriteableRegistry();
+            this.stateSupplier = () -> readCompressed(version, bytes, namedWriteableRegistry);
+            this.refCounted = bytes;
+        } else {
+            // older versions just contain the bare state
+            final var state = ClusterState.readFrom(in, null);
+            this.stateSupplier = () -> state;
+            this.refCounted = null;
+        }
+    }
+
+    private static ClusterState readCompressed(Version version, BytesReference bytes, NamedWriteableRegistry namedWriteableRegistry)
+        throws IOException {
+        final var compressor = CompressorFactory.compressor(bytes);
+        StreamInput in = bytes.streamInput();
+        try {
+            if (compressor != null) {
+                in = new InputStreamStreamInput(compressor.threadLocalInputStream(in));
+            }
+            in = new NamedWriteableAwareStreamInput(in, namedWriteableRegistry);
+            in.setVersion(version);
+            try (StreamInput input = in) {
+                return ClusterState.readFrom(input, null);
+            }
+        } finally {
+            IOUtils.close(in);
```

Comment:
I believe `in` might be already closed by line `try (StreamInput input = in)` in case when no exception. I wonder if it is possible to simplify code to have a single try with resources instead nested one

Reply:
Yes I think you're right, I tidied this up in f9df3b96af498949661e235e5b8822fdfb576a82. This comes from the code in `PublicationTransportHandler` which dates back a long way (b80324db084bd42bcb36457286157e3c26ae435d and beyond) and has accumulated a bunch of cruft over the years.